cpp/Contact.cpp

```diff
@@ -549,6 +549,202 @@ void dolfinx_contact::Contact::assemble_vector(
+  // correct map
+  std::shared_ptr<const dolfinx::graph::AdjacencyList<int>> map
+      = _facet_maps[pair];
+  const std::vector<xt::xtensor<double, 2>>& qp_phys = _qp_phys[puppet_mt];
+  const std::vector<std::pair<std::int32_t, int>>& puppet_facets
+      = _cell_facet_pairs[puppet_mt];
+  std::shared_ptr<const dolfinx::graph::AdjacencyList<int>> facet_map
+      = submesh.facet_map();
+  const std::size_t max_links
+      = *std::max_element(_max_links.begin(), _max_links.end());
+  const std::size_t num_facets = puppet_facets.size();
+  const std::size_t num_q_points
+      = _quadrature_rule->offset()[1] - _quadrature_rule->offset()[0];
+
+  xt::xtensor<double, 2> q_points
+      = xt::zeros<double>({std::size_t(num_q_points), std::size_t(gdim)});
+
+  std::vector<std::int32_t> perm(num_q_points);
+  std::vector<std::int32_t> linked_cells(num_q_points);
+
+  // Create output vector
+  std::vector<PetscScalar> c(
+      num_facets * num_q_points * max_links * ndofs * gdim, 0.0);
+  const auto cstride = int(num_q_points * max_links * ndofs * gdim);
+
+  // Loop over all facets
+  for (std::size_t i = 0; i < num_facets; i++)
+  {
+    const tcb::span<const int> links = map->links((int)i);
+    assert(links.size() == num_q_points);
+
+    // Compute Pi(x) form points x and gap funtion Pi(x) - x
+    for (std::size_t j = 0; j < num_q_points; j++)
+    {
+      const tcb::span<const int> linked_pair = facet_map->links(links[j]);
+      linked_cells[j] = linked_pair[0];
+      const std::size_t row = i * num_q_points;
+      for (std::size_t k = 0; k < gdim; k++)
+        q_points(j, k) = qp_phys[i](j, k) + gap[row * gdim + j * gdim + k]
+                         - u_packed[row * gdim + j * gdim + k];
+    }
+    // Sort linked cells
+    assert(linked_cells.size() == num_q_points);
+    const auto [unique_cells, offsets] = dolfinx_contact::sort_cells(
+        xtl::span(linked_cells.data(), linked_cells.size()),
+        xtl::span(perm.data(), perm.size()));
+
+    // Loop over sorted array of unique cells
+    for (std::size_t j = 0; j < unique_cells.size(); ++j)
+    {
+
+      std::int32_t linked_cell = parent_cells[unique_cells[j]];
+      // Extract indices of all occurances of cell in the unsorted cell
+      // array
+      auto indices
+          = xtl::span(perm.data() + offsets[j], offsets[j + 1] - offsets[j]);
+
+      // Extract all physical points Pi(x) on a facet of linked_cell
+      auto qp = xt::view(q_points, xt::keep(indices), xt::all());
+      // Compute values of basis functions for all y = Pi(x) in qp
+      std::array<std::size_t, 4> b_shape
+          = evaluate_basis_shape(*_V, indices.size(), 1);
+      if (b_shape[3] > 1)
+        throw std::invalid_argument(
+            "pack_test_functions assumes values size 1");
+      xt::xtensor<double, 4> basis_values(b_shape);
+      std::fill(basis_values.begin(), basis_values.end(), 0);
+      std::vector<std::int32_t> cells(indices.size(), linked_cell);
```

Comment:
Could we move this declaration outside (with max number of cells/number of q points) and resize it instead?